## **Handling Missing Values**

In [29]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [30]:
df = pd.read_csv('datasets/synthetic_dataset.csv')

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  1614 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    2312 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


##### There are some missing values

##### The second way of finding whether we have null values in the data is by using the isnull() function.

In [32]:
print(df.isnull().sum())

Category    2748
Price        174
Rating      2050
Stock       1352
Discount     392
dtype: int64


## Handling Missing Values

***1. Deleting the columns with missing data***

In [33]:
updated_df = df.dropna(axis=1)

In [34]:
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Empty DataFrame


##### The problem with this method is that we may lose valuable information on that feature, as we have deleted it completely due to some null values.

##### Should only be used if there are too many null values.

***2. Deleting the rows with missing data***

In [35]:
updated_df = df.dropna(axis=0)

In [36]:
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 540 entries, 6 to 4343
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  540 non-null    object 
 1   Price     540 non-null    float64
 2   Rating    540 non-null    float64
 3   Stock     540 non-null    object 
 4   Discount  540 non-null    float64
dtypes: float64(3), object(2)
memory usage: 25.3+ KB


##### In this case, there are possibilities of getting better accuracy than before. This might be because the columns contains more valuable information than we expected.

***3. Filling the Missing Values – Imputation***

In this case, we will be filling the missing values with a certain number.
The possible ways to do this are:

- Filling the missing data with the mean or median value if it’s a numerical variable.
- Filling the missing data with mode if it’s a categorical value.
- Filling the numerical value with 0 or -999, or some other number that will not occur in the data. This can be done so that the machine can recognize that the data is not real or is different.
- Filling the categorical value with a new type for the missing values.

In [37]:
df['Rating'].mean()

np.float64(3.0382925191086185)

In [38]:
df['Rating'].median()

np.float64(3.0820600518841306)

In [39]:
#fillna: fills the null records
#dropna: drops the null records

updated_df = df
updated_df['Rating']=updated_df['Rating'].fillna(df['Rating'].mean())
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  1614 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [40]:
updated_df1 = df
updated_df1['Rating']=updated_df['Rating'].fillna(df['Rating'].median())
updated_df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  1614 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [41]:
updated_df = df
updated_df['Price']=updated_df['Price'].fillna(df['Price'].mean())
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  1614 non-null   object 
 1   Price     4362 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [42]:
df['Category'].mode()

0    C
Name: Category, dtype: object

In [43]:
df['Category'].mode()[0]

'C'

In [44]:
updated_df = df
updated_df['Category'] = updated_df['Category'].fillna(updated_df['Category'].mode()[0])
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4362 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [46]:
updated_df = df
updated_df['Stock'] = updated_df['Stock'].fillna(updated_df['Stock'].mode()[0])
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4362 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     4362 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [47]:
updated_df = df
updated_df['Discount'] = updated_df['Discount'].fillna(updated_df['Discount'].median())
updated_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4362 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     4362 non-null   object 
 4   Discount  4362 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


***4. Forward & Backward Filling – Imputation***

In [49]:
df = pd.read_csv('datasets/synthetic_dataset.csv')

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  1614 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    2312 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [51]:
df1 = df

In [61]:
#Backwards fill
df1['Category'] = df1['Category'].bfill()

In [60]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    2312 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [54]:
df1.tail()

,Category,Price,Rating,Stock,Discount
4357,B,4436.0,4.728335,NaN,49.0
4358,B,6236.0,NaN,Out of Stock,4.0
4359,D,3283.0,NaN,Out of Stock,9.0
4360,D,2999.0,4.425995,NaN,40.0
4361,NaN,NaN,2.184471,NaN,0.0


##### Last row is NaN that's why we got 4361 values

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4361 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    2312 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


In [56]:
df = pd.read_csv('datasets/synthetic_dataset.csv')

In [62]:
#forward fill
df1['Category'] = df1['Category'].ffill()

In [58]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4188 non-null   float64
 2   Rating    2312 non-null   float64
 3   Stock     3010 non-null   object 
 4   Discount  3970 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB


##### Filling all values with forward and backward filling

In [63]:
df1 = df1.bfill().ffill()

In [64]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4362 entries, 0 to 4361
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Category  4362 non-null   object 
 1   Price     4362 non-null   float64
 2   Rating    4362 non-null   float64
 3   Stock     4362 non-null   object 
 4   Discount  4362 non-null   float64
dtypes: float64(3), object(2)
memory usage: 170.5+ KB
